File: src/Kattis/Programming Challenges/Assignments/A6/F.cpp
#include <iostream>

using namespace std;

int main()
{
    // Get number of squares
    int destination;
    cin >> destination;

    // Get cost per square
    int prices[destination];
    for (int i = 0; i < destination; i++)
    {
        cin >> prices[i];
    }

    // Initialize dp matrix with values = -1
    int dp[destination + 1][destination + 1];
    for (int i = 0; i < destination + 1; i++)
    {
        for (int j = 0; j < destination + 1; j++)
        {
            dp[i][j] = -1;
        }
    }

    // DP backwards
    int step, square, forward, backward, value;
    for (step = destination; step >= 0; step--)
    {
        for (square = 1; square < destination + 1; square++)
        {
            forward = square + step + 1;
            backward = square - step;

            // Check if square is destination
            if (square == destination)
            {
                dp[step][square] = prices[destination - 1];
            }
            
            // Check if we can move forward
            if (forward <= destination && step < destination)
            {
                if (dp[step + 1][forward] != -1)
                {
                    dp[step][square] = dp[step + 1][forward] + prices[square - 1];
                }
            }

            // Check if we can move backward
            if (backward > 0)
            {
                if (dp[step][backward] != -1)
                {
                    value = dp[step][backward] + prices[square - 1];
                    if (dp[step][square] == -1 || value < dp[step][square])
                    {
                        dp[step][square] = value;
                    }
                }
            }
        }
    }

    // Output results
    cout << dp[1][2] << endl;

    // Successful exit
    return 0;
}

File: src/Kattis/Programming Challenges/Assignments/A6/F.py
# Get input
dest = int(input())
prices = [int(input()) for _ in range(dest)]
dp = [[-1] * (dest + 1) for _ in range(dest + 1)] # Dynamic programming matrix

# Start from the end
for step in range(dest, -1, -1):
    for square in range (1, dest + 1):
        if square == dest:
            dp[step][square] = prices[-1]
            continue
        square_forward = square + step + 1
        square_backward = square - step
        if square_forward <= dest and step < dest:
            if dp[step + 1][square_forward] != -1:
                dp[step][square] = dp[step + 1][square_forward] + prices[square - 1]
        if square_backward > 0:
            if dp[step][square_backward] != -1:
                value = dp[step][square_backward] + prices[square - 1]
                if dp[step][square] == -1 or value < dp[step][square]:
                    dp[step][square] = value
print(dp[1][2])
